### PR TITLE
WIP Reduce redundant state changes.

### PR DESCRIPTION
**Pull request by roboman2444**
WIP, but its probably better if someone with more knowledge of the engine finished it...

Still needs texture, framebuffer, buffer state management.

Works on #524

### CODE REVIEW

**Comment by roboman2444** [comment]
I'm gonna need to know the max texture units in order to do no redundant texture bindings.

**Comment by yvt** [comment]
The water renderer uses up to 16 texture units, which is the maximum number of texture units supported by macOS's OpenGL compatible profile driver (no matter how many units are supported by the hardware).

**Comment by roboman2444** [comment]
Is there any performance or quality improvement that can be made by using more texture units if the hardware/software supports it?

**Comment by yvt** [comment]
While making the level 3 water renderer (`r_water 3`) I had to drop one dynamic water texture because of 16 texture units limitation, and as a consequence the repeating pattern is more visible than that of the level 2 renderer. Also, the level 2 and 3 water renderers use a special version of a lighting shader named `CommonNoSSAO` to reduce the texture unit usage.

Except that, the limitation has never been a problem.

**Comment by roboman2444** [comment]
Have you thought about possibly using texture arrays for that?
You also only use 2 components from wavetexture2 and wavetexture3. You could possibly pack them into a single 4 component texture.

**Comment by yvt** [comment]
Oh, why didn't I notice that? Thanks for pointing that out!

File: Sources/Gui/SDLGLDevice.h [hunk 1]
```diff
@@ -25,16 +25,54 @@
...
+			float lw_w;
+			//DepthFunc
+			spades::draw::IGLDevice::Enum df_func;
+			//UseProgram
+			spades::draw::IGLDevice::UInteger up_program;
+			//Texture stuff
+			unsigned char tex_active;
+			//change the number if you want more texture units!
+			spades::draw::IGLDevice::UInteger tex_id[SPADES_SDLGL_TEXTURE_UNITS];
+			spades::draw::IGLDevice::Enum tex_target[SPADES_SDLGL_TEXTURE_UNITS];
```
Comment (yvt) [inline Sources/Gui/SDLGLDevice.h]: I'm no expert of OpenGL, but it seems that OpenGL doesn't work like that:

https://www.khronos.org/opengl/wiki/Texture#Texture_image_units

> Each texture image unit supports bindings to all targets. So a 2D texture and an array texture can be bound to the same image unit, or different 2D textures can be bound in two different image units without affecting each other. So which texture gets used when rendering? In GLSL, this depends on the type of sampler that uses this texture image unit.

**Comment by roboman2444** [comment]
For buffers im gonna need more information on how buffers are used in AOS.
I may also suggest using VAOs. It might give a slight performance improvement, and it is required to have a VAO bound to conform to 3.3 core spec.

**Comment by yvt** [comment]
> For buffers im gonna need more information on how buffers are used in AOS.

All buffer objects are merely plain old VBOs containing either vertex data or indices. There's no read backs nor feedbacks.

> I may also suggest using VAOs. It might give a slight performance improvement, and it is required to have a VAO bound to conform to 3.3 core spec.

Okay, I'll give it a try.

**Comment by yvt** [comment]
It turned out that we can't use VAOs on macOS, unless we switch to the core profile and port all shaders to GLSL 1.50 or later, but doing so will force us to drop the support for pre-OpenGL 3.2 hardware.

**Pull request closed without merging**